### PR TITLE
fix: remove conversation member (AR-2065)

### PR DESCRIPTION
**Pull request by Garzas**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

There was some parsing issues so I've changed some models

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt [hunk 1]
```diff
@@ -394,15 +396,21 @@ class ConversationDataSource(
...
                         Either.Left(it)
                     }, {
-                        wrapStorageRequest {
-                            conversationDAO.deleteMemberByQualifiedID(
-                                idMapper.toDaoModel(userID),
-                                idMapper.toDaoModel(conversationId)
-                            )
+                        when (it) {
+                            // TODO I think on unchanged we should also remove member from local storage
+                            // probably there was some synchronisation issue that this user was already removed
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:405]: Yes, we should!

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/conversation/ConversationApiImpl.kt [hunk 2]
```diff
@@ -77,16 +69,38 @@ class ConversationApiImpl internal constructor(private val authenticatedNetworkC
...
+    override suspend fun removeMember(
+        userId: UserId,
+        conversationId: ConversationId
+    ): NetworkResponse<ConversationMemberChangeDTO> = try {
+        httpClient.delete(
+            "$PATH_CONVERSATIONS/${conversationId.domain}/${conversationId.value}/$PATH_MEMBERS/${userId.domain}/${userId.value}"
+        ).let { response ->
+            when (response.status) {
+                HttpStatusCode.OK -> wrapKaliumResponse<ConversationMemberChangeDTO.Removed> { response }
+                // TODO should we show user some message that there was no change?
```
Comment (typfel) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/conversation/ConversationApiImpl.kt:102]: I'd say no, this is a race condition when two users perform the same action which resolves it self so the user doesn't need to retry his action.

**Review by typfel — CHANGES_REQUESTED** [review]
Looks good! Just need to also remove/add the user locally on `NoChange`.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt [hunk 3]
```diff
@@ -394,15 +396,21 @@ class ConversationDataSource(
...
+                            ConversationMemberChangeDTO.Unchanged -> Either.Right(Unit)
+                            is ConversationMemberChangeDTO.Added -> wrapStorageRequest {
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:407]: how come a delete request could return with a member added response?

Comment (Garzas) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt]: Refactored to two undependable models

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/conversation/ConversationMemberChangeDTO.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        @SerialName("qualified_from")
+        val fromUser: UserId,
+        @SerialName("time")
+        val time: String
+    ) : ConversationMemberAddedDTO()
+}
+
+@Serializable
+sealed class ConversationMemberRemovedDTO {
+    // TODO: the server response with an event aka, UserAdded model is inaccurate
```
Comment (mchenani) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/conversation/ConversationMemberChangeDTO.kt:37]: I think this comment should be updated aswell.

**Review by mchenani — APPROVED** [review]
LGTM✅

**Review by typfel — APPROVED** [review]

**Pull request merged**